src/irr/core/CFixedCapacityDoublyLinkedList.h

```diff
@@ -0,0 +1,117 @@
+#ifndef __DOUBLY_LINKED_LIST_H_INCLUDED__
+#define __DOUBLY_LINKED_LIST_H_INCLUDED__
+#include "irr/core/alloc/PoolAddressAllocator.h"
+#include "irr/core/Types.h"
+
+namespace irr {
+	namespace core {
+
+		template<typename Value>
+		struct Snode
+		{
+			Value data;
+			Snode* prev;
+			Snode* next;
+
+			Snode(const Value& val) : data(val)
+			{ }
+			Snode(Snode&& other) : data(std::move(other.data)), prev(std::move(other.prev)), next(std::move(other.next))
+			{ }
+
+			Snode& operator=(const Snode& other)
+			{
+				this.data = other.data;
+				this.prev = other.prev;
+				this.next = other.next;
+				return *this;
+			}
+
+			Snode& operator=(Snode&& other)
+			{
+				this.data = std::move(other.data);
+				this.prev = std::move(other.prev);
+				this.next = std::move(other.next);
+				return *this;
+			}
+		};
+
+		template<typename Value>
+		class DoublyLinkedList
+		{
+			typedef Snode<Value>* iteratorptr_t;
+
+		private:
+			void* reservedSpace;
+			PoolAddressAllocator<uint32_t> alloc;
+			iteratorptr_t p_begin;
+			iteratorptr_t p_end;
+			uint32_t cap;
+
+			inline void popBack()
+			{
+				if(p_end->prev != nullptr)
+					p_end->prev->next = nullptr;
+				iteratorptr_t temp = p_end;
+				p_end = p_end->prev;
+				alloc.free_addr(reinterpret_cast<uint32_t>(temp), 1u);
+			}
+
+		public:
+
+			inline void pushFront(Value &val) 
+			{
+
+				uint32_t addr = alloc.alloc_addr(1u, 1u);
+				iteratorptr_t n = new(reinterpret_cast<iteratorptr_t>(reservedSpace) + addr) Snode(val);
```

Comment:
no that's still not how it works!

reserved space is for the ALLOCATOR INTERNAL STATE.

You need to create a separate array of `SNode[cap]`